### DISCUSSION TITLE
Reading HRLDAS output in R

### DISCUSSION THREAD

**Discussion by fdesales** [NCAR/hrldas#222] [General]
Hi. Any recommendations on how to read HRLDAS output files in R?  I am able to read WRF output files using the `eixport` package, however, HRLDAS output files do not contain latitude/longitude (e.g., XLAT_M/XLONG_M, etc) grid information. Thanks!

**Comment by cenlinhe** [comment]
@RAbolafiaRosenzweig Can you help with this since you are familiar with R?

**Reply by fdesales** [reply]
Sure. How can I help?

**Reply by RAbolafiaRosenzweig** [reply]
I first read in model outputs from the LDASOUT files, and then the latitude-longitude grid from the HRLDAS_setup file. However, my scripts for this use MATLAB (with the ncread command).

**Reply by PGhaneei** [reply]
Hi,

This is an example of how to read and plot outputs from LDASOUT, as well as the grid information from the HRLDAS_setup file in R:


#### Install packages
install.packages("ncdf4")     
install.packages("reshape2")  
install.packages("ggplot2")   

library(ncdf4)
library(ggplot2)
library(reshape2)

#### Read output
file_path <- "./2015040300.LDASOUT_DOMAIN1"
nc <- nc_open(file_path)

#### Extract the soil moisture 
smc <- ncvar_get(nc, "SOIL_M")

nc_close(nc)

#### Select the first layer of sil moisture:
smc_slice <- smc[ , 1, ]

smc_melt <- melt(smc_slice)

#### Plot 
ggplot(smc_melt, aes(Var1, Var2, fill = value)) +
  geom_raster() +
  scale_fill_gradientn(colors = terrain.colors(10), limits = c(0, 1)) +
  theme_minimal() +
  labs(title = "Soil Moisture (SOIL_M)", x = "Longitude Index", y = "Latitude Index", fill = "SOIL_M") +
  coord_fixed()


#### Lat and Lon of grids 

file_path_grids <- "./HRLDAS_setup_2015040100_d1"
nc_grids <- nc_open(file_path_grids)

#### Extract the XLAT 
XLAT <- ncvar_get(nc_grids, "XLAT")

#### Extract the XLONG 
XLONG <- ncvar_get(nc_grids, "XLONG")

**Reply by fdesales** [reply]
Thanks @PGhaneei for your script.  I was wondering if your ggplot map is correctly projected?  In my case, I can get ggplot to generate a map but its geographical projection is wrong.  Would you know how to use XLAT and XLONG to reproject the map?

**Reply by fdesales** [reply]
For example, this is what I get from ggplot using your script.  Lat/longs are incorrect.  My domain should be center over Botswana.  Having the correct projection is crucial for calculating spatial averages and plotting geo boundaries.  Thanks
![Rplot01](https://github.com/user-attachments/assets/1d022f11-808b-41a2-a49b-d8662aed1bf8)

**Reply by fdesales** [reply]
I have been able to fix the lat/lon issue using the ext function. Please see below.   I still to make sure the projection is correct, though.  I am using the R terra package to handle raster data more easily.

```
#Read latitude and longitude from the setup file (forcing)
grd = rast("HRLDAS_setup_2020090100_d1")
xlat = grd[[1]]
xlon = grd[[2]]

#Make a domain with max and min lats and lons
ymn = min(values(xlat))
ymx = max(values(xlat))
xmn = min(values(xlon))
xmx = max(values(xlon))
domain = ext(xmn,xmx,ymn,ymx)

#Read model output and list variable names
dat = rast("2020090500.LDASOUT_DOMAIN1")
names(dat)

#Extract the first variable (IVGTYP) and fix lat/lon
vtype = dat[[1]]
ext(vtype) <- domain  #correct lat/lon
```

**Reply by PGhaneei** [reply]
For spatial analysis, what I do is create a new data frame with XLAT and XLONG. Here is an example:

### Install necessary packages
install.packages("ncdf4")     
install.packages("reshape2")  
install.packages("ggplot2")   
install.packages("spatstat")
install.packages("sf")
install.packages("sp")

library(ncdf4)
library(ggplot2)
library(reshape2)
library(spatstat)
library(sf)
library(sp)

### Read output
file_path <- "./2015040300.LDASOUT_DOMAIN1"
nc <- nc_open(file_path)

### Extract the soil moisture 
smc <- ncvar_get(nc, "SOIL_M")

nc_close(nc)

### Select the first layer of soil moisture:
smc_slice <- smc[ , 1, ]

############## Lat and Lon of grids ############
file_path_grids <- "./HRLDAS_setup_2015040100_d1"
nc_grids <- nc_open(file_path_grids)

### Extract the XLAT 
XLAT <- ncvar_get(nc_grids, "XLAT")

### Extract the XLONG 
XLONG <- ncvar_get(nc_grids, "XLONG")

nc_close(nc_grids)

############## Create Complete Data Frame for Spatial Analysis ############

spatial_data <- data.frame(
  lon = as.vector(XLONG),
  lat = as.vector(XLAT),
  soil_m = as.vector(smc_slice)
)

############## Convert to Geo Data Frames ############
spatial_data_sf <- st_as_sf(spatial_data, coords = c("lon", "lat"), crs = 4326)


### Plot
ggplot(data = spatial_data_sf) +
  geom_sf(aes(color = soil_m)) +
  scale_color_viridis_c( limits = c(0, 1)) +
  labs(title = "soil_m",
       color = "soil_m") +
  theme_minimal()

**Comment by FuYingWang** [comment]
There is a package named 'rwrfhydro' in R.
![image](https://github.com/user-attachments/assets/72821f50-0ecb-4995-b1da-bb2510e58ef8)

**Reply by FuYingWang** [reply]
For multi-layer variables such as' SOIL_M',  'inLyr=1' is also required to extract the value of a certain layer

**Reply by fdesales** [reply]
I tried using the `rwrfhydro `package but it did not work for HRLDAS output files.

**Comment by FuYingWang** [comment]
Thanks @PGhaneei , it help a lot. Is there a way to output 'spatial_data_stf' in tiff format?

**Comment by fdesales** [comment]
Hi all. I am still stuck on this issue of how to reproject HRLDAS output files into a geographical latlon projection correctly.  To do this correctly, I need to know what actual projection the HRLDAS files are on.  A check with cdo gives me the project is "generic", 

```
cdo sinfo 2020103121.LDASOUT_DOMAIN1
   Grid coordinates :
     1 : generic                  : points=121801 (349x349)
     2 : generic                  : points=1047 (349x3)
     3 : generic                  : points=1396 (349x4)
```

But ncdump shows "map_proj = 3" which indicates a Mercator map projection in WRF.  Very confusing!

                :MAP_PROJ = 3 ;    
                :LAT1 = -26.88026f ;
                :LON1 = 19.40043f ;
                :DX = 3000.f ;
                :DY = 3000.f ;
                :TRUELAT1 = -22.268f ;
                :TRUELAT2 = -999.9f ;
                :STAND_LON = -999.9f ;

Question: Can anyone recommend an approach to reproject from Mercator to longlat?  I have tried the rwrfhydro library in R but it only works for Lambert and Conformal Conic projections according to the manual. I can use CDO or NCL.
